crates/common/consensus/src/misc.rs

```diff
@@ -0,0 +1,158 @@
+use std::cmp::max;
+
+use alloy_primitives::{aliases::B32, B256};
+use anyhow::ensure;
+use ethereum_hashing::{hash, Context, Sha256Context};
+use tree_hash::Hash256;
+
+use crate::{
+    deneb::beacon_state::BeaconState,
+    fork_choice::helpers::constants::{
+        MAX_EFFECTIVE_BALANCE, MAX_SEED_LOOKAHEAD, SHUFFLE_ROUND_COUNT, SLOTS_PER_EPOCH,
+    },
+    fork_data::ForkData,
+};
+
+static MAX_RANDOM_BYTE: u64 = 255;
+
+pub fn compute_shuffled_index(
+    index: usize,
+    index_count: usize,
+    seed: B256,
+) -> anyhow::Result<usize> {
+    ensure!(index < index_count, "Index must be less than index_count");
+
+    let mut index = index;
+    for round in 0..SHUFFLE_ROUND_COUNT {
+        let pivot = bytes_to_int64(&hash_with_round(seed, round)[..]) % index_count as u64;
+        index = do_round(seed, index, pivot as usize, round, index_count);
+    }
+    Ok(index)
+}
+
+fn hash_with_round(seed: B256, round: u8) -> Hash256 {
+    let mut context = Context::new();
+
+    context.update(&seed.0);
+    context.update(&[round]);
+
+    let digest = context.finalize();
+    Hash256::from_slice(digest.as_ref())
+}
+
+fn bytes_to_int64(slice: &[u8]) -> u64 {
+    let mut bytes = [0; 8];
+    bytes.copy_from_slice(&slice[0..8]);
+    u64::from_le_bytes(bytes)
+}
+
+fn do_round(seed: B256, index: usize, pivot: usize, round: u8, list_size: usize) -> usize {
+    let flip = (pivot + (list_size - index)) % list_size;
+    let position = max(index, flip);
+    let source = hash_with_round_and_position(seed, round, position);
+    let byte_index = (position % 256) / 8;
+    let byte = source.get(byte_index).unwrap();
+    let bit = (byte >> (position % 8)) & 1;
+
+    if bit == 1 {
+        flip
+    } else {
+        index
+    }
+}
+
+fn hash_with_round_and_position(seed: B256, round: u8, position: usize) -> Hash256 {
+    let mut context = Context::new();
+
+    context.update(seed.as_slice());
+    context.update(&[round]);
+
+    context.update(&(position / 256).to_le_bytes()[0..4]);
+
+    let digest = context.finalize();
+    Hash256::from_slice(digest.as_ref())
+}
```

Comment:
```suggestion
pub fn compute_shuffled_index(
    index: usize,
    index_count: usize,
    seed: B256,
) -> anyhow::Result<usize> {
    ensure!(index < index_count, "Index must be less than index_count");

    let mut index = index;
    for round in 0..SHUFFLE_ROUND_COUNT {
        let seed_with_round = [seed.as_slice(), &round.to_le_bytes())[..]].concat();
        let pivot = bytes_to_int64(&hash(&seed_with_round)[..]) % index_count as u64;
        
        let flip = (pivot + (list_size - index)) % list_size;
        let position = max(index, flip);
        let seed_with_position = [seed_with_round.as_slice(), &(position / 256).to_le_bytes()].concat()
        let source = hash(&seed_with_position);
        let byte = source[(position % 256) / 8];
        let bit = (byte >> (position % 8)) % 2;
  
          index = if bit == 1 {
              flip
          } else {
              index
          };
    }
    Ok(index)
}

fn bytes_to_int64(slice: &[u8]) -> u64 {
    let mut bytes = [0; 8];
    bytes.copy_from_slice(&slice[0..8]);
    u64::from_le_bytes(bytes)
}
```

I think this is a lot cleaner